### PR TITLE
#167727627 Set up Sign Up Validations

### PR DESCRIPTION
**Pull request by ChidozieGabriel**
#### What does this PR do?
Set up sign up validations

#### Description of Task to be completed?
Return error with status code of 400 when user signs up with:
- Already existing email address
- An invalid email address
- A password of length less than 8
- A non-alphanumeric password

#### How should this be manually tested?
* Clone the repo and `cd` into the project directory
* Run `yarn install` to install the project's dependency
* Run `yarn test` to start the app

#### What are the relevant pivotal tracker stories?
[#167727627](https://www.pivotaltracker.com/story/show/167727627)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -5,11 +5,12 @@
...
     "coverage": "nyc report --reporter=text-lcov | coveralls",
-    "test": "set NODE_ENV=test&& nyc mocha --require @babel/register 'src/**/**.test.js' --no-timeout ",
-    "dev-start": "nodemon --exec babel-node src/index.js",
+    "pretest": "NODE_ENV=test yarn drop && yarn migrate",
+    "test": "NODE_ENV=test nyc mocha --recursive --exit --require @babel/register --require @babel/polyfill 'src/**/**.test.js' --no-timeout",
+    "dev-start": "NODE_ENV=development nodemon --exec babel-node src/index.js",
     "build": "babel src --out-dir build",
     "start": "node build/index.js",
-    "drop": "sequelize-cli db:migrate:undo:all",
+    "drop": "./node_modules/.bin/babel-node ./node_modules/.bin/sequelize db:migrate:undo:all",
```
Comment (Daymorelah) [inline package.json]: Name this script `undo-migration` (it is more descriptive) instead of `drop`

File: package.json [hunk 2]
```diff
@@ -5,11 +5,12 @@
   "main": "index.js",
   "scripts": {
     "coverage": "nyc report --reporter=text-lcov | coveralls",
-    "test": "set NODE_ENV=test&& nyc mocha --require @babel/register 'src/**/**.test.js' --no-timeout ",
-    "dev-start": "nodemon --exec babel-node src/index.js",
+    "pretest": "NODE_ENV=test yarn drop && yarn migrate",
```
Comment (Daymorelah) [inline package.json]: there is no need for the `pretest` script. add the content of this script to the `test` script

File: package.json [hunk 3]
```diff
@@ -38,20 +39,21 @@
...
     "@babel/node": "^7.5.5",
+    "@babel/polyfill": "^7.4.4",
     "@babel/preset-env": "^7.5.5",
     "@babel/register": "^7.5.5",
     "chai": "^4.2.0",
     "chai-http": "^4.3.0",
+    "coveralls": "^3.0.6",
+    "eslint": "^5.16.0",
+    "eslint-config-airbnb-base": "^14.0.0",
+    "eslint-config-prettier": "^6.0.0",
```
Comment (Daymorelah) [inline package.json:54]: Why did you install prettier?

File: src/index.js [hunk 4]
```diff
@@ -4,6 +4,8 @@ import bodyParser from 'body-parser';
 import cors from 'cors';
 import swaggerUI from 'swagger-ui-express';
 import doc from '../doc.json';
+import routesV1 from './routes/v1';
```
Comment (Daymorelah) [inline src/index.js]: use `routes` instead of `routesV1`

File: src/index.js [hunk 5]
```diff
@@ -14,11 +16,15 @@ app.use(cors());
 // api doc
 app.use('/api-docs', swaggerUI.serve, swaggerUI.setup(doc));
 
+app.use('/api/v1', routesV1);
+
 app.get('/', (req, res) => res.status(200).send({ message: 'Welcome to Barefoot Nomad' }));
 app.all('*', (req, res) => res.send({ message: 'route not found' }));
 
-app.listen(port, () => {
+models.sequelize.sync().then(() => {
```
Comment (Daymorelah) [inline src/index.js]: Why are you calling sequelize in the index.js file? anything sequelize should be done in the controller's or model's folder

File: package.json (hunk 3 above)
Comment (ChidozieGabriel) [reply package.json:54]: @Daymorelah I did not install prettier. I got it when I rebased my branch with the develop branch.
Here is the [package.json](https://github.com/andela/rogue-backend/blob/develop/package.json) _(line 45)_ for the develop branch.

File: src/index.js (hunk 5 above)
Comment (ChidozieGabriel) [reply src/index.js]: I've resolved it. Thanks.

File: src/index.js (hunk 4 above)
Comment (ChidozieGabriel) [reply src/index.js]: I've resolved it. Thanks.

File: package.json (hunk 2 above)
Comment (ChidozieGabriel) [reply package.json]: I've resolved it. Thanks.

File: package.json (hunk 1 above)
Comment (ChidozieGabriel) [reply package.json]: I've resolved it. Thanks.

File: src/models/User.js [hunk 6]
```diff
@@ -1,22 +1,66 @@
...
+            args: true,
+            msg: 'Email address field cannot be empty.'
+          },
+          async exists(value) {
+            const user = await User.findOne({
+              where: { email: value }
+            });
+
+            if (user) {
+              throw new Error('Email address already exists.');
```
Comment (Daymorelah) [inline src/models/User.js]: Due to security reasons, we must not tell the user that the email already exist. `User details already exist` is much better and safer

Comment (Daymorelah) [inline src/models/User.js]: Due to security reasons, we must not tell the user that the email already exist. `User details already exist` is much better and safer

File: src/models/User.js [hunk 7]
```diff
@@ -1,22 +1,66 @@
...
+        unique: true,
+        allowNull: false,
+        validate: {
+          isEmail: {
+            args: true,
+            msg: 'Email address is invalid.'
+          },
+          notNull: {
+            args: true,
+            msg: 'Email address field cannot be empty.'
```
Comment (Daymorelah) [inline src/models/User.js]: Good job @ChidozieGabriel. However, I feel `Email address cannot be empty.` is more suitable.

File: src/models/User.js (hunk 6 above)
Comment (Daymorelah) [reply src/models/User.js]: Setting `unique` to true already caters for the code on line 20-30. I like the way you are able to show that we can use a custom function for validation.

File: src/models/User.js [hunk 8]
```diff
@@ -1,22 +1,66 @@
...
+          }
+        }
+      },
+      password: {
+        type: DataTypes.STRING,
+        allowNull: false,
+        validate: {
+          len: {
+            args: [8],
+            msg: 'Password must be at least 8 characters long.'
```
Comment (Daymorelah) [inline src/models/User.js]: minimum of 6 characters will suffice

File: src/models/User.js (hunk 6 above)
Comment (Daymorelah) [reply src/models/User.js]: So you can set validation on the `unique` property and pass the error message i suggested above. i.e. `{ unique: { args: true, msg: 'User details already exist' } }`

File: src/models/index.js [hunk 9]
```diff
@@ -18,9 +18,8 @@ if (env !== 'production') {
   sequelize = new Sequelize(process.env.DATABASE_URL);
 }
 
-fs
-  .readdirSync(__dirname)
-  .filter((file) => (file.indexOf('.') !== 0) && (file !== basename) && (file.slice(-3) === '.js'))
+fs.readdirSync(__dirname)
+  .filter((file) => file.indexOf('.') !== 0 && file !== basename && file.slice(-3) === '.js')
```
Comment (Daymorelah) [inline src/models/index.js]: Remove the parenthesis around  the `file` argument on line 22 so that hound can make us rest

File: src/models/index.js [hunk 10]
```diff
@@ -34,5 +33,6 @@ Object.keys(db).forEach((modelName) => {
 
 db.sequelize = sequelize;
 db.Sequelize = Sequelize;
+db.sequelize.sync();
```
Comment (Daymorelah) [inline src/models/index.js]: why did you add line 36?. What were you trying to achieve here?. Please, shed more light on this.

**Review by Daymorelah — CHANGES_REQUESTED** [review]

File: src/models/index.js (hunk 9 above)
Comment (ChidozieGabriel) [reply src/models/index.js]: :-)

File: src/models/index.js (hunk 10 above)
Comment (ChidozieGabriel) [reply src/models/index.js]: `db.sequelize.sync()` creates tables that do not exist. Therefore running `yarn dev-start` returns error because there are no tables. 
A better approach could be to call `yarn migrate` in the `dev-start` script in package.json thus:
```
{
  dev-start: yarn migrate && ...
}
```
and then delete the line 36 `db.sequelize.sync()`.

Please could you tell your opinion? What do I do?

File: src/models/User.js (hunk 6 above)
Comment (ChidozieGabriel) [reply src/models/User.js]: I tried it but it did not work. That is why I chose to use a custom validation for it.
I think I should delete line 9, 
```
{
  ...
  unique: true,
  ...
}
```
seeing that it is redundant.

Comment (ChidozieGabriel) [reply src/models/User.js]: I've resolved it. Thanks.

File: src/models/User.js (hunk 7 above)
Comment (ChidozieGabriel) [reply src/models/User.js]: I've resolved it. Thanks.

File: src/models/index.js (hunk 9 above)
Comment (ChidozieGabriel) [reply src/models/index.js]: I've resolved it. Thanks.

File: src/models/User.js (hunk 8 above)
Comment (ChidozieGabriel) [reply src/models/User.js]: I've resolved it. Thanks.

File: src/models/index.js [hunk 11]
```diff
@@ -7,20 +7,21 @@ import sequelizeConfig from '../config/config';
...
   sequelize = new Sequelize(database, username, password, config);
 } else {
   sequelize = new Sequelize(process.env.DATABASE_URL);
 }
 
-fs
-  .readdirSync(__dirname)
-  .filter((file) => (file.indexOf('.') !== 0) && (file !== basename) && (file.slice(-3) === '.js'))
+fs.readdirSync(__dirname)
+  .filter((file) => file.indexOf('.') !== 0 && file !== basename && file.slice(-3) === '.js')
```
[the remark this answers is not in the document]
Comment (Daymorelah) [reply src/models/index.js]: resolve this hound error

File: src/models/index.js (hunk 9 above)
Comment (Daymorelah) [reply src/models/index.js]: I cant see it being implemented. Take a look at line 22

File: src/test/integrationTest/authRoute.test.js [hunk 12]
```diff
@@ -0,0 +1,94 @@
+/* eslint-disable camelcase */
+import chai, { expect, should } from 'chai';
+import chaiHTTP from 'chai-http';
+import server from '../..';
+
+should();
+chai.use(chaiHTTP);
+const route = '/api/v1/auth/signup';
+const generateNumber = () => `${Math.round(Math.random() * 1000)}${Date.now()}`;
```
Comment (Daymorelah) [inline src/test/integrationTest/authRoute.test.js]: This line of code is not necessary.

File: src/test/integrationTest/authRoute.test.js [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+};
+
+const nullEmail = {
+  password: 'noNeedForPasswordAnyMore'
+};
+
+const nullPassword = {
+  email: `john${generateNumber()}@doe.com`
+};
+
```
Comment (Daymorelah) [inline src/test/integrationTest/authRoute.test.js]: Instead of creating different request body in here, you can always seed the database. However, seeding the database for this test is not necessary. Since this is an integration test, just create different if blocks that would test each of your validation. Also, try to use `expect` instead of `should` for your assertions.

File: src/test/integrationTest/authRoute.test.js [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+      const { status } = res.body;
+      expect(status).to.eql(res.status);
+      done();
+    })
+    .catch(done);
+};
+
+describe('USER VALIDATION TEST', () => {
+  before('Register a valid user', (done) => {
+    chai
```
Comment (Daymorelah) [inline src/test/integrationTest/authRoute.test.js]: you should not be registering a valid user in your test. if your tests needs data to be in the database beforehand, then seed the database before running your test.

File: src/test/integrationTest/authRoute.test.js [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+  });
+
+  it('should throw error 400 when user signs up with a null password', (done) => {
+    runTest(done, nullPassword);
+  });
+
+  it('should throw error 400 when user signs up with a null email and null password', (done) => {
+    runTest(done);
+  });
+});
```
Comment (Daymorelah) [inline src/test/integrationTest/authRoute.test.js]: you would need to re-write your tests because I don't understand the tests you wrote.

**Review by Daymorelah — CHANGES_REQUESTED** [review]

File: src/models/index.js (hunk 9 above)
Comment (ChidozieGabriel) [reply src/models/index.js]: Okay, I'm sorry.
I've resolved it now. Thanks.

File: src/test/integrationTest/authRoute.test.js (hunk 12 above)
Comment (ChidozieGabriel) [reply src/test/integrationTest/authRoute.test.js]: I've resolved it. Thanks.

File: src/test/integrationTest/authRoute.test.js (hunk 14 above)
Comment (ChidozieGabriel) [reply src/test/integrationTest/authRoute.test.js]: I've resolved it. Thanks.

File: src/test/integrationTest/authRoute.test.js (hunk 15 above)
Comment (ChidozieGabriel) [reply src/test/integrationTest/authRoute.test.js]: Since all my tests are the same code, instead of repeating it, I encapsulated it in a function.
Line 37
```
const runTest = (done, object = {}) => {
  chai
    .request(server)
    .post(route)
    .send(object)
    .then(res => {
      expect(res).to.have.status(400);
      const { body } = res;
      expect(body).to.be.an('object');
      expect(body).to.have.property('error');
      const { status } = body;
      expect(status).to.eql(res.status);
      done();
    })
    .catch(done);
};
```

File: src/test/integrationTest/authRoute.test.js (hunk 13 above)
Comment (ChidozieGabriel) [reply src/test/integrationTest/authRoute.test.js]: I've resolved it. Thanks.
